src/document-context/measure-custom-elements.ts

```diff
@@ -0,0 +1,382 @@
+    const boundRegisterElement: typeof originalRegisterElement =
+        originalRegisterElement.bind(document);
+    function wrappedRegisterElement(
+        tagName: string, options: RegisterElementOptions) {
+      const maybeProto = options.prototype;
+      if (!maybeProto) {
+        return boundRegisterElement(tagName, options);
+      }
+      const proto = maybeProto;
+      let elementCounter = 0;
+      const originalCreate = proto.createdCallback || (() => undefined);
+
+      function wrap(name: string, fullName: string) {
+        const original = proto[fullName] || (() => undefined);
+        proto[fullName] = function(this: any) {
+          const counter: number = this[idSymbol];
+
+          return makeMeasurement(
+              name, tagName, counter, () => original.apply(this, arguments));
+        };
+      }
+
+      proto.createdCallback = function(this: any) {
+        const counter = elementCounter++;
+        this[idSymbol] = counter
+
+        const result = makeMeasurement(
+            'created', tagName, counter, () => originalCreate.apply(this));
+
+        const originalPropertySetter = this._propertySetter;
+        this._propertySetter = function(this: any) {
+          return makeMeasurement(
+              'data', tagName, counter,
+              () => originalPropertySetter.apply(this, arguments));
+        };
+        const originalNotifyPath = this.notifyPath;
+        this.notifyPath = function(this: any) {
+          return makeMeasurement(
+              'data', tagName, counter,
+              () => originalNotifyPath.apply(this, arguments));
+        };
+
+        return result;
+      };
+      wrap('connected', 'attachedCallback');
+      wrap('disconnected', 'detachedCallback');
+      wrap('attributeChanged', 'attributeChangedCallback');
+
+      return makeMeasurement(
+          'registered', tagName, null,
+          () => boundRegisterElement(tagName, options));
+    }
+    document.registerElement = wrappedRegisterElement;
+  }
+
+  if (window.customElements) {
+    const originalDefine = window.customElements.define;
+    const boundDefine: typeof originalDefine =
+        originalDefine.bind(window.customElements);
+    let elementCounter = 0;
+
+    function wrappedDefineElement(
+        tagName: string,
+        constructor: CustomElementV1Constructor): CustomElementV1Constructor {
+      let wrappedConstructor: CustomElementV1Constructor =
+          class extends constructor {
+        constructor() {
+          const counter = elementCounter++;
+          const startMark = `${prefix}start created ${tagName} ${counter}`;
+          const endMark = `${prefix}end created ${tagName} ${counter}`;
+          const measure = `${prefix}created ${tagName} ${counter}`;
+
+          window.performance.mark(startMark);
+          try {
+            super();
+          } finally {
+            window.performance.mark(endMark);
+            window.performance.measure(measure, startMark, endMark);
```

Comment:
no clear way to ensure that it gets called before the timeline is closed when doing that asynchronously